File: donkserv/internal/skill/skill.go
package skill

import (
	"os"
	"path/filepath"
	"time"
)

// Metadata Skill元数据
// 遵循 Claude Code / Open Agent Skills 规范
type Metadata struct {
	Name                   string            `yaml:"name"`                     // 技能名称（小写字母、数字、短横线，最多64字符）
	Description            string            `yaml:"description"`              // 技能描述及使用时机
	Version                string            `yaml:"version"`                  // 版本号
	Author                 string            `yaml:"author"`                   // 作者
	Homepage               string            `yaml:"homepage"`                 // 主页URL
	Tags                   []string          `yaml:"tags"`                     // 标签（用于分类）
	ArgumentHint           string            `yaml:"argument-hint"`            // 自动补全参数提示
	DisableModelInvocation bool              `yaml:"disable-model-invocation"` // 禁止自动触发，仅能手动调用
	UserInvocable          bool              `yaml:"user-invocable"`           // 是否可用户调用（斜杠命令）
	AllowedTools           []string          `yaml:"allowed-tools"`            // 允许使用的工具列表
	License                string            `yaml:"license"`                  // 许可证信息
	Compatibility          string            `yaml:"compatibility"`            // 环境要求
	Metadata               map[string]string `yaml:"metadata"`                 // 自定义键值对
}

// RuntimeConfig Skill运行时配置
type RuntimeConfig struct {
	Requires []string `yaml:"requires"` // 依赖的其他技能
	Examples []string `yaml:"examples"` // 使用示例
}

// SkillDirs Skill包含的目录
type SkillDirs struct {
	Scripts    string // scripts/ 目录路径
	References string // references/ 目录路径
	Assets     string // assets/ 目录路径
}

// Skill Skill核心结构
// 遵循 Claude Code Agent Skills 三层渐进式披露架构
type Skill struct {
	metadata     Metadata      // 元数据（始终加载到system prompt）
	runtime      RuntimeConfig // 运行时配置
	instructions string        // 核心指令（触发时加载）
	dirs         SkillDirs     // 子目录路径
	baseDir      string        // 技能根目录
	name         string        // 目录名称
	loaded       time.Time     // 加载时间
}

// NewSkill 创建新的Skill实例
// 参数:
//   - name: 技能目录名称
//   - description: 技能描述
//
// 返回:
//   - *Skill: Skill实例
func NewSkill(name, description string) *Skill {
	return &Skill{
		name: name,
		metadata: Metadata{
			Name:                   name,
			Description:            description,
			Version:                "1.0.0",
			UserInvocable:          true,
			DisableModelInvocation: false,
		},
		runtime: RuntimeConfig{},
		dirs:    SkillDirs{},
		loaded:  time.Now(),
	}
}

// Name 获取技能名称
// 参数:
//   - 无
//
// 返回:
//   - string: 技能名称
func (s *Skill) Name() string {
	if s.metadata.Name != "" {
		return s.metadata.Name
	}
	return s.name
}

// Description 获取技能描述
// 参数:
//   - 无
//
// 返回:
//   - string: 技能描述
func (s *Skill) Description() string {
	return s.metadata.Description
}

// Version 获取版本号
// 参数:
//   - 无
//
// 返回:
//   - string: 版本号
func (s *Skill) Version() string {
	return s.metadata.Version
}

// Author 获取作者
// 参数:
//   - 无
//
// 返回:
//   - string: 作者
func (s *Skill) Author() string {
	return s.metadata.Author
}

// Homepage 获取主页URL
// 参数:
//   - 无
//
// 返回:
//   - string: 主页URL
func (s *Skill) Homepage() string {
	return s.metadata.Homepage
}

// Tags 获取标签列表
// 参数:
//   - 无
//
// 返回:
//   - []string: 标签列表
func (s *Skill) Tags() []string {
	return s.metadata.Tags
}

// ArgumentHint 获取参数提示
// 参数:
//   - 无
//
// 返回:
//   - string: 参数提示
func (s *Skill) ArgumentHint() string {
	return s.metadata.ArgumentHint
}

// IsUserInvocable 是否可用户调用
// 参数:
//   - 无
//
// 返回:
//   - bool: 是否可用户调用
func (s *Skill) IsUserInvocable() bool {
	return s.metadata.UserInvocable
}

// DisableModelInvocation 是否禁用自动触发
// 参数:
//   - 无
//
// 返回:
//   - bool: 是否禁用自动触发
func (s *Skill) DisableModelInvocation() bool {
	return s.metadata.DisableModelInvocation
}

// AllowedTools 获取允许的工具列表
// 参数:
//   - 无
//
// 返回:
//   - []string: 允许的工具列表
func (s *Skill) AllowedTools() []string {
	return s.metadata.AllowedTools
}

// License 获取许可证信息
// 参数:
//   - 无
//
// 返回:
//   - string: 许可证信息
func (s *Skill) License() string {
	return s.metadata.License
}

// Compatibility 获取环境要求
// 参数:
//   - 无
//
// 返回:
//   - string: 环境要求
func (s *Skill) Compatibility() string {
	return s.metadata.Compatibility
}

// CustomMetadata 获取自定义元数据
// 参数:
//   - 无
//
// 返回:
//   - map[string]string: 自定义键值对
func (s *Skill) CustomMetadata() map[string]string {
	return s.metadata.Metadata
}

// GetCustomMetadata 获取指定的自定义元数据值
// 参数:
//   - key: 键名
//
// 返回:
//   - string: 元数据值，如果不存在则返回空字符串
func (s *Skill) GetCustomMetadata(key string) string {
	if s.metadata.Metadata == nil {
		return ""
	}
	return s.metadata.Metadata[key]
}

// Requires 获取依赖列表
// 参数:
//   - 无
//
// 返回:
//   - []string: 依赖的技能列表
func (s *Skill) Requires() []string {
	return s.runtime.Requires
}

// Examples 获取使用示例
// 参数:
//   - 无
//
// 返回:
//   - []string: 使用示例列表
func (s *Skill) Examples() []string {
	return s.runtime.Examples
}

// Instructions 获取核心指令
// 参数:
//   - 无
//
// 返回:
//   - string: 核心指令内容
func (s *Skill) Instructions() string {
	return s.instructions
}

// BaseDir 获取技能根目录
// 参数:
//   - 无
//
// 返回:
//   - string: 根目录路径
func (s *Skill) BaseDir() string {
	return s.baseDir
}

// Dirs 获取子目录信息
// 参数:
//   - 无
//
// 返回:
//   - SkillDirs: 子目录路径
func (s *Skill) Dirs() SkillDirs {
	return s.dirs
}

// Loaded 获取加载时间
// 参数:
//   - 无
//
// 返回:
//   - time.Time: 加载时间
func (s *Skill) Loaded() time.Time {
	return s.loaded
}

// SetMetadata 设置元数据
// 参数:
//   - m: Metadata实例
func (s *Skill) SetMetadata(m Metadata) {
	s.metadata = m
}

// SetRuntime 设置运行时配置
// 参数:
//   - r: RuntimeConfig实例
func (s *Skill) SetRuntime(r RuntimeConfig) {
	s.runtime = r
}

// SetInstructions 设置核心指令
// 参数:
//   - instructions: 指令内容
func (s *Skill) SetInstructions(instructions string) {
	s.instructions = instructions
}

// SetBaseDir 设置根目录
// 参数:
//   - baseDir: 根目录路径
func (s *Skill) SetBaseDir(baseDir string) {
	s.baseDir = baseDir
	s.dirs = SkillDirs{
		Scripts:    filepath.Join(baseDir, "scripts"),
		References: filepath.Join(baseDir, "references"),
		Assets:     filepath.Join(baseDir, "assets"),
	}
}

// SetName 设置目录名称
// 参数:
//   - name: 目录名称
func (s *Skill) SetName(name string) {
	s.name = name
}

// GetDirPath 获取子目录路径
// 参数:
//   - dirType: 目录类型 (scripts, references, assets)
//
// 返回:
//   - string: 目录路径
func (s *Skill) GetDirPath(dirType string) string {
	switch dirType {
	case "scripts":
		return s.dirs.Scripts
	case "references":
		return s.dirs.References
	case "assets":
		return s.dirs.Assets
	default:
		return ""
	}
}

// HasScripts 检查是否有scripts目录
// 参数:
//   - 无
//
// 返回:
//   - bool: 是否存在
func (s *Skill) HasScripts() bool {
	info, err := os.Stat(s.dirs.Scripts)
	return err == nil && info.IsDir()
}

// HasReferences 检查是否有references目录
// 参数:
//   - 无
//
// 返回:
//   - bool: 是否存在
func (s *Skill) HasReferences() bool {
	info, err := os.Stat(s.dirs.References)
	return err == nil && info.IsDir()
}

// HasAssets 检查是否有assets目录
// 参数:
//   - 无
//
// 返回:
//   - bool: 是否存在
func (s *Skill) HasAssets() bool {
	info, err := os.Stat(s.dirs.Assets)
	return err == nil && info.IsDir()
}

// ListScripts 列出scripts目录下的文件
// 参数:
//   - 无
//
// 返回:
//   - []string: 文件路径列表
func (s *Skill) ListScripts() ([]string, error) {
	return s.listDirFiles("scripts")
}

// ListReferences 列出references目录下的文件
// 参数:
//   - 无
//
// 返回:
//   - []string: 文件路径列表
func (s *Skill) ListReferences() ([]string, error) {
	return s.listDirFiles("references")
}

// ListAssets 列出assets目录下的文件
// 参数:
//   - 无
//
// 返回:
//   - []string: 文件路径列表
func (s *Skill) ListAssets() ([]string, error) {
	return s.listDirFiles("assets")
}

// listDirFiles 列出指定目录下的所有文件
// 参数:
//   - dirType: 目录类型
//
// 返回:
//   - []string: 文件路径列表
//   - error: 错误信息
func (s *Skill) listDirFiles(dirType string) ([]string, error) {
	dirPath := s.GetDirPath(dirType)
	if dirPath == "" {
		return nil, nil
	}

	info, err := os.Stat(dirPath)
	if err != nil {
		return nil, nil
	}

	if !info.IsDir() {
		return nil, nil
	}

	entries, err := os.ReadDir(dirPath)
	if err != nil {
		return nil, err
	}

	var files []string
	for _, entry := range entries {
		if !entry.IsDir() {
			files = append(files, filepath.Join(dirPath, entry.Name()))
		}
	}

	return files, nil
}

// GetDirName 获取目录名称
// 参数:
//   - 无
//
// 返回:
//   - string: 目录名称
func (s *Skill) GetDirName() string {
	return s.name
}

File: donkui/lib/common/model/skill.dart
/// Skill 数据模型
/// 用于表示 Skill 的基本信息
class Skill {
  /// Skill 名称（唯一标识）
  final String name;

  /// Skill 描述
  final String description;

  /// 版本号
  final String version;

  /// 作者
  final String author;

  /// 标签列表
  final List<String> tags;

  /// 是否启用
  final bool enabled;

  /// 是否允许用户通过斜杠命令调用
  final bool userInvocable;

  /// 是否禁止自动触发
  final bool disableModelInvocation;

  /// Skill 目录路径
  final String path;

  /// 是否包含 scripts 目录
  final bool hasScripts;

  /// 是否包含 references 目录
  final bool hasReferences;

  /// 是否包含 assets 目录
  final bool hasAssets;

  Skill({
    required this.name,
    required this.description,
    required this.version,
    required this.author,
    required this.tags,
    required this.enabled,
    required this.userInvocable,
    required this.disableModelInvocation,
    required this.path,
    required this.hasScripts,
    required this.hasReferences,
    required this.hasAssets,
  });

  /// 从 JSON 数据创建 Skill 实例
  factory Skill.fromJson(Map<String, dynamic> json) {
    return Skill(
      name: json['name'] as String? ?? '',
      description: json['description'] as String? ?? '',
      version: json['version'] as String? ?? '',
      author: json['author'] as String? ?? '',
      tags:
          (json['tags'] as List<dynamic>?)
              ?.map((e) => e as String? ?? '')
              .toList() ??
          [],
      enabled: json['enabled'] as bool? ?? false,
      userInvocable: json['user_invocable'] as bool? ?? false,
      disableModelInvocation:
          json['disable_model_invocation'] as bool? ?? false,
      path: json['path'] as String? ?? '',
      hasScripts: json['has_scripts'] as bool? ?? false,
      hasReferences: json['has_references'] as bool? ?? false,
      hasAssets: json['has_assets'] as bool? ?? false,
    );
  }

  /// 将 Skill 实例转换为 JSON 数据
  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
      'version': version,
      'author': author,
      'tags': tags,
      'enabled': enabled,
      'user_invocable': userInvocable,
      'disable_model_invocation': disableModelInvocation,
      'path': path,
      'has_scripts': hasScripts,
      'has_references': hasReferences,
      'has_assets': hasAssets,
    };
  }

  /// 复制当前实例并修改指定字段
  Skill copyWith({
    String? name,
    String? description,
    String? version,
    String? author,
    List<String>? tags,
    bool? enabled,
    bool? userInvocable,
    bool? disableModelInvocation,
    String? path,
    bool? hasScripts,
    bool? hasReferences,
    bool? hasAssets,
  }) {
    return Skill(
      name: name ?? this.name,
      description: description ?? this.description,
      version: version ?? this.version,
      author: author ?? this.author,
      tags: tags ?? this.tags,
      enabled: enabled ?? this.enabled,
      userInvocable: userInvocable ?? this.userInvocable,
      disableModelInvocation:
          disableModelInvocation ?? this.disableModelInvocation,
      path: path ?? this.path,
      hasScripts: hasScripts ?? this.hasScripts,
      hasReferences: hasReferences ?? this.hasReferences,
      hasAssets: hasAssets ?? this.hasAssets,
    );
  }
}

/// Skill 列表响应模型
class SkillListResponse {
  /// Skill 列表
  final List<Skill> data;

  /// 总数
  final int total;

  SkillListResponse({required this.data, required this.total});

  /// 从 JSON 数据创建 SkillListResponse 实例
  factory SkillListResponse.fromJson(Map<String, dynamic> json) {
    return SkillListResponse(
      data:
          (json['data'] as List<dynamic>)
              .map((e) => Skill.fromJson(e as Map<String, dynamic>))
              .toList(),
      total: json['total'] as int,
    );
  }
}

/// Skill 指令响应模型
class SkillInstructionsResponse {
  /// 指令内容（Markdown 格式）
  final String instructions;

  SkillInstructionsResponse({required this.instructions});

  /// 从 JSON 数据创建 SkillInstructionsResponse 实例
  factory SkillInstructionsResponse.fromJson(Map<String, dynamic> json) {
    return SkillInstructionsResponse(
      instructions: json['instructions'] as String,
    );
  }
}

/// Skill 脚本列表响应模型
class SkillScriptsResponse {
  /// 脚本文件列表
  final List<String> scripts;

  SkillScriptsResponse({required this.scripts});

  /// 从 JSON 数据创建 SkillScriptsResponse 实例
  factory SkillScriptsResponse.fromJson(Map<String, dynamic> json) {
    return SkillScriptsResponse(
      scripts:
          (json['scripts'] as List<dynamic>).map((e) => e as String).toList(),
    );
  }
}

/// Skill 脚本内容响应模型
class SkillScriptContentResponse {
  /// 脚本名称
  final String name;

  /// 脚本内容
  final String content;

  SkillScriptContentResponse({required this.name, required this.content});

  /// 从 JSON 数据创建 SkillScriptContentResponse 实例
  factory SkillScriptContentResponse.fromJson(Map<String, dynamic> json) {
    return SkillScriptContentResponse(
      name: json['name'] as String,
      content: json['content'] as String,
    );
  }
}

/// 操作响应模型
class SkillActionResponse {
  /// 响应消息
  final String message;

  SkillActionResponse({required this.message});

  /// 从 JSON 数据创建 SkillActionResponse 实例
  factory SkillActionResponse.fromJson(Map<String, dynamic> json) {
    return SkillActionResponse(message: json['message'] as String);
  }
}

/// Skill 错误响应模型
class SkillErrorResponse {
  /// 错误信息
  final String error;

  SkillErrorResponse({required this.error});

  /// 从 JSON 数据创建 SkillErrorResponse 实例
  factory SkillErrorResponse.fromJson(Map<String, dynamic> json) {
    return SkillErrorResponse(error: json['error'] as String);
  }
}
